20241014-wordcloud/src/routes/+page.svelte

```diff
@@ -0,0 +1,170 @@
+<script lang="ts">
+    import { onMount } from "svelte";
+    import WordCloud from "./WordCloud.svelte";
+
+    let words = [
+        {text: "placeholder", count: 1},
+    ];
+    let submission: string = "";
+    let submissionInput: HTMLInputElement;
+    const maxTimer = 5;
+
+    function invalidate(placeholder: string) {
+        submissionInput.placeholder = placeholder;
+        submissionInput.style.transition = "";
+        submissionInput.value = "";
+        submissionInput.style.backgroundColor = "#ffccdb";
+        setTimeout(() => {
+            submissionInput.style.transition = "background-color 0.5s";
+            submissionInput.style.backgroundColor = "white";
+        }, 100);
+        setTimeout(() => {
+            submissionInput.placeholder = "Enter a word...";
+        }, 600);
+    }
+
+    async function submitWord() {
+        const rgx = /drop\s+table/i;
+        if (rgx.test(submission)) {
+            alert("Nice try, but my SQL inputs are thoroughly sanitised.");
+        }
+
+        if (submission === "") {
+            return;
+        }
+
+        // Sanitise input
+        submission = submission
+            .toLowerCase()
+            .trim()
+            .split(/\s+/).join("-")
+            .replace(/[^a-z0-9-_.]/g, "");
+
+        if (submission === "") {
+            invalidate("invalid character in input");
+            return;
+        }
+
+        if (localStorage.getItem(`wordcloud___${submission}`) === "true") {
+            invalidate("already submitted");
+            return;
+        } else {
+            localStorage.setItem(`wordcloud___${submission}`, "true");
+        }
+
+        const response = await fetch("/answer", {
+            method: "POST",
+            headers: {
+                "Content-Type": "application/json",
+            },
+            body: JSON.stringify({text: submission}),
+        });
+        words = await response.json();
+        timer = maxTimer;
+        submission = "";
+    }
+
+    async function getWords() {
+        const response = await fetch("/ask");
+        words = await response.json();
+        // Print to console so that it's easy to get hold of the data if we need to replot
+        let s = [];
+        for (const {text, count} of words) {
+            for (let i = 0; i < count; i++) {
+                s.push(text);
+            }
+        }
+        console.log(s.join(" "));
+    }
+
+    let timer = maxTimer;
+
+    onMount(async () => {
+        await getWords();
+        setInterval(async () => {
+            timer -= 1;
+            if (timer <= 0) {
+                await getWords();
+                timer = maxTimer;
+            }
+        }, 1000);
+    });
```

Comment:
So I was trying around a bit (without really knowing what I'm doing) with getting automatic updates from the database. One could set up an SSE connection here:
```

    onMount(async () => {
        await getWords();
        
        // Set up SSE connection
        const eventSource = new EventSource('/updates');
        
        eventSource.onmessage = (event) => {
            words = JSON.parse(event.data);
        };
        
        // Clean up 
        return () => {
            eventSource.close();
        };
    });
```

And then have a new endpoint `src/routes/updates/+server.ts`:

```
import { getData } from '$lib/server/db';
import type { RequestHandler } from './$types';

export const GET: RequestHandler = async () => {
    const headers = new Headers({
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache',
        'Connection': 'keep-alive'
    });

    const stream = new ReadableStream({
        async start(controller) {
            while (true) {
                const data = await getData("programming");
                controller.enqueue(`data: ${JSON.stringify(data)}\n\n`);
                // Wait a bit before next check
                await new Promise(resolve => setTimeout(resolve, 1000));
            }
        }
    });

    return new Response(stream, { headers });
```

I don't know enough web to know whether this is actually better and scales, but as I understand it, the difference here is basically that we update once there's something new, but we also have one long-lived connection per user and I'm not sure whether this is good.

Reply:
I also came across [WebSockets](https://medium.com/@dev-john-nguyen/how-i-use-websockets-to-make-updates-in-real-time-e15b726f2b80), but seems a bit overkill here.

Reply:
I really like your thinking here @mastoffel. Keeping a connection open does seem like it's going to hinder scaling for an application like this, where the user isn't particularly concerned about immediate feedback. But server-side events does look it was made for a case like this.

In fact I really don't really think there's anything wrong with the polling as it is, but it's probably preferable to poll for an update flag ("time since last update on the server") rather than polling for all of the data every time.